### DISCUSSION TITLE
Issue with Maplibre URL-encoding

### DISCUSSION THREAD

**Discussion by Z3vron** [maplibre/maplibre-native#2706] [Q&A] [answered]
Hi guys
I am new to MapLibre so my issue might be silly for you but I can't fixed it and I spend already some time on it.
I am writing simple application that is using Maplibre with Mapy.cz API. I can launch application and it shows first layer good but when i try to change it than it gives me error 404. I looked into Logcat and it looks for me that Maplibre is URL-encoding  values : 
[HTTP] Request with response = 404: No additional information
[HTTP] This request was cancelled (https://api.mapy.cz/v1/maptiles/outdoor/256/%7Bz%7D/%7Bx%7D/%7By%7D?apikey =my_api_key

If I try to open that in the web browser it doesn't work but if I replace few values (https://api.mapy.cz/v1/maptiles/outdoor/256/1/1/1?apikey=my_api_key) than it works normally. From my understanding Maplibre automaticlly is URL-encoding values into format from the first link but I don't know why.  I tried using openstreetmap and it ended with same issues. 

This is part of the code that I think is responsible for the issue

` private fun initializeMapLayer(selectedLayer: String) {
        mapLibreMap.getStyle { style ->
            removeExistingLayerAndSource(style)

            val urlTemplate = "https://api.mapy.cz/v1/maptiles/$selectedLayer/256/{z}/{x}/{y}?apikey=$API_KEY"
            Log.d("MapDebug", "URL before adding to source: $urlTemplate")

            val rasterSource = RasterSource("raster-tiles", urlTemplate, 256)
            style.addSource(rasterSource)

            val rasterLayer = RasterLayer("raster-layer", "raster-tiles")
            style.addLayer(rasterLayer)

            Log.d("MapDebug", "Layer added with source: raster-tiles")
        }
    }

private fun removeExistingLayerAndSource(style: Style) {
        style.getLayer("raster-layer")?.let { layer ->
            Log.d("MapDebug", "Removing existing layer: raster-layer")
            style.removeLayer(layer)
        }

        style.getSource("raster-tiles")?.let { source ->
            Log.d("MapDebug", "Removing existing source: raster-tiles")
            style.removeSource(source)
        }
    }`
Log cat: 
Removing existing layer: raster-layer
Removing existing source: raster-tiles
URL before adding to source: https://api.mapy.cz/v1/maptiles/outdoor/256/{z}/{x}/{y}apikey=my_api_key
Layer added with source: raster-tiles

From my understanding it should put values into z/x/y so that it could render tiles but it url-encodes them first. 

PS: my_api_key I put in the logcat logs where my real Api key is so that it stayes private.
I am using android-sdk:11.0.1 
If it is needed I can upload my project so that yo ucan see more code/configuration but it is in absolute chaos right now as I am trying different things to make it to work

**Comment by louwers** [comment, marked as answer]
I would recommend using a vector-tile source with MapLibre Native.

But if you are set on using mapy.cz, you can also use their TileJSON endpoint. Try setting this as style:

```json
{
  "version": 8,
  "name": "Empty Style",
  "metadata": {"maputnik:renderer": "mlgljs"},
  "sources": {
    "mapy": {
      "type": "raster",
      "url": "https://api.mapy.cz/v1/maptiles/outdoor/tiles.json?apikey=your_api_key"
    }
  },
  "sprite": "",
  "glyphs": "https://orangemug.github.io/font-glyphs/glyphs/{fontstack}/{range}.pbf",
  "layers": [{"id": "", "type": "raster", "source": "mapy"}],
  "id": "srots50"
}
```

**Reply by Z3vron** [reply]
Thanks for an answear, reading your solution I remembered that I seen something similiar and I realize that I already used that style and it works for the original layer when app is starting but later during changing layers I used url from their TileJSON end point directly and that's why it wasn't working..
